clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -354,6 +358,460 @@ fatbinary(ArrayRef<std::pair<StringRef, StringRef>> InputFiles,
+        continue;
+      }
+      if (Line.empty())
+        return createStringError(inconvertibleErrorCode(),
+                                 "Invalid SYCL Table file.");
+      auto [FirstWord, Rem1] = Line.split("|");
+      SYCLTableEntry Entry;
+      Entry.IRFile = FirstWord.str();
+      if (Rem1.empty())
+        return createStringError(inconvertibleErrorCode(),
+                                 "Invalid SYCL Table file.");
+      auto [SecondWord, ThirdWord] = Rem1.split("|");
+      Entry.PropFile = SecondWord.str();
+      Entry.SymFile = ThirdWord.str();
+      Entries.push_back(Entry);
+    }
+    return Error::success();
+  }
+};
+
+// Run LLVM to SPIR-V translation.
+static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef InputTable,
+                                                     const ArgList &Args) {
+  Table LiveSYCLTable;
+  Expected<std::string> LLVMToSPIRVPath =
+      findProgram("llvm-spirv", {getMainExecutable("llvm-spirv")});
+  if (!LLVMToSPIRVPath)
+    return LLVMToSPIRVPath.takeError();
+
+  if (Error Err = LiveSYCLTable.populateSYCLTable(InputTable))
+    return std::move(Err);
+  auto InputFiles = LiveSYCLTable.getListOfIRFiles();
+
+  SmallVector<StringRef, 8> CmdArgs;
+  CmdArgs.push_back(*LLVMToSPIRVPath);
+  StringRef LLVMToSPIRVOptions;
+  if (Arg *A = Args.getLastArg(OPT_llvm_spirv_options_EQ))
+    LLVMToSPIRVOptions = A->getValue();
+  LLVMToSPIRVOptions.split(CmdArgs, " ", /* MaxSplit = */ -1,
+                           /* KeepEmpty = */ false);
+  CmdArgs.push_back("-o");
+
+  int I = 0;
+  for (const auto &File : InputFiles) {
+    // Create a new file to write the translated file to.
+    auto TempFileOrErr =
+        createOutputFile(sys::path::filename(ExecutableName), "spv");
+    if (!TempFileOrErr)
+      return TempFileOrErr.takeError();
+    BumpPtrAllocator Alloc;
+    StringSaver Saver(Alloc);
+
+    CmdArgs.push_back(*TempFileOrErr);
+    CmdArgs.push_back(File);
+    if (Error Err = executeCommands(*LLVMToSPIRVPath, CmdArgs))
+      return std::move(Err);
+    // Replace bc file in SYCL table with spv file
+    LiveSYCLTable.Entries[I++].IRFile = *TempFileOrErr;
+    // Pop back last two items
+    CmdArgs.pop_back_n(2);
+  }
+  auto Output = LiveSYCLTable.writeSYCLTableToFile();
+  if (!Output)
+    return Output.takeError();
+  return *Output;
+}
+
+// Run clang-offload-wrapper and llc
+static Expected<StringRef> runWrapper(StringRef &InputFile,
+                                      const ArgList &Args) {
+  // Create a new file to write the wrapped file to.
+  auto TempFileOrErr =
+      createOutputFile(sys::path::filename(ExecutableName), "bc");
+  if (!TempFileOrErr)
+    return TempFileOrErr.takeError();
+  // Create a new file to write the output of llc to.
+  auto OutputFileOrErr =
+      createOutputFile(sys::path::filename(ExecutableName), "o");
+  if (!OutputFileOrErr)
+    return OutputFileOrErr.takeError();
+
+  // call to clang-offload-wrapper
+  {
```

Comment:
Nit: If we need to use {} for scoping we probably should move those to subfunctions

Reply:
I am trying to following the community flow here. In the community code, we call wrap and llc in one common routine.

Thanks

Reply:
Thinking again, I decided to move to subfunctions. Thanks